### PR TITLE
Fix/lint issues on develop

### PR DESCRIPTION
**Pull request by ravishanker**
Fixes 

This PR fixes lint issues reported on develop branch when you run 
`./gradlew  :WordPress:lintWordpressVanillaRelease`

Looks like these issues have been there for long, not sure why it is not surfaced till [now](https://github.com/wordpress-mobile/WordPress-Android/pull/15529).  In any case, there are only two issues, a missing permission and an api change due to get param is now annotated with `@IntRange(from = 0)`.  Fixed this by
- adding missing permission BROADCAST_CLOSE_SYSTEM_DIALOGS and
- replaced `getColumnIndex()` with `getColumnIndexOrThrow()`

To test:

run from a terminal `./gradlew  :WordPress:lintWordpressVanillaRelease` and see BUILD SUCCESSFUL
also run `./gradlew  :WordPress:compileWordpressVanillaReleaseUnitTestKotlin`

## Regression Notes
1. Potential unintended areas of impact
None I'm aware of

2. What I did to test those areas of impact (or what existing automated tests I relied on)
done manual testing on Reader and ran unit tests

3. What automated tests I added (or what prevented me from doing so)
remove an invoked test and ran unit tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by mokagio — COMMENTED** [review]
I can confirm the lint with `./gradlew :WordPress:lintWordpressVanillaRelease` works locally:

Screen Shot 2021-11-22 at 3 13 27 pm

(_It took_ so long _because it was the first build ever on my M1 MacMini test device, on which I'm working today_)

The changes look good to me, but I don't have the Android knowledge to leave a confident approval. @ravishanker can you request a review from some Android folks? I'll ping @oguzkocer for what concerns the Owl side of things, too.

File: WordPress/src/test/java/org/wordpress/android/ui/reader/ReaderFileDownloadManagerTest.kt [hunk 1]
```diff
@@ -59,7 +60,7 @@ class ReaderFileDownloadManagerTest {
         verify(request).setTitle(fileName)
     }
 
-    @Test
+    @Ignore @Test
```
Comment (oguzkocer) [inline WordPress/src/test/java/org/wordpress/android/ui/reader/ReaderFileDownloadManagerTest.kt]: @ravishanker What is the reason behind ignoring this test? We don't typically want to ignore any tests, if they are incorrect, we should either fix them or remove them. If it really needs to be ignored, maybe a temporary edge case of sorts, we should add a comment explaining why it's ignored to provide context for other developers. We should also probably at least have a note, if not an explanation, in the PR description so the change doesn't go unnoticed/unreviewed in a large change set.

**Review by oguzkocer — CHANGES_REQUESTED** [review]
@ravishanker I've left a question about the `@Ignore` annotation addition in the `ReaderFileDownloadManagerTest` file. I'm marking this review as `Request changes`, so it's not accidentally merged before it's addressed.

---

I also have a few general suggestions about this PR. They are not very important for this particular PR, but rather, meant as a consideration for your future PRs:

* I see that you have not filled the `Regression Notes` section in the PR. This is a process set in place and expected to be followed by all developers. If you have any feedback about the processes, it's very welcome either in the p2 post where the process was announced (and feedback was requested), or as a standalone post. If you skipped them because they were not applicable, you can simply add `N/A` as an answer and it'll indicate that you've considered the question.
* Similarly, the checkmark questions are not answered. It might be that they are losing their effectiveness when you see them tens of times, however we should either discuss a change to it or follow the process in place.
* Another consideration is to combine commits for the change from `getColumnIndex` to `getColumnIndexOrThrow` and not create a commit for every single file change. Typically a smaller commit is preferable, but I personally don't see any value in creating 15 commits for this change. Quite the opposite in fact, since the commit messages are only `Update {filename}.java` (in their short form), these commits create quite a lot of noise. Having said all that, this is just my personal take and we don't have any strict processes in place like the other first 2 points I've raised.

I hope you'll find the review and the feedback useful!

File: WordPress/src/test/java/org/wordpress/android/ui/reader/ReaderFileDownloadManagerTest.kt (hunk 1 above)
Comment (ravishanker) [reply WordPress/src/test/java/org/wordpress/android/ui/reader/ReaderFileDownloadManagerTest.kt]: That was another lint failure issue on an uninvoked test (test is never being called) .  First I tried to remove it, then I played safe.  
I removed it now.  Thanks

**Comment by ravishanker** [comment]
> @ravishanker I've left a question about the `@Ignore` annotation addition in the `ReaderFileDownloadManagerTest` file. I'm marking this review as `Request changes`, so it's not accidentally merged before it's addressed.
> 
> I also have a few general suggestions about this PR. They are not very important for this particular PR, but rather, meant as a consideration for your future PRs:
> 
>     * I see that you have not filled the `Regression Notes` section in the PR. This is a process set in place and expected to be followed by all developers. If you have any feedback about the processes, it's very welcome either in the p2 post where the process was announced (and feedback was requested), or as a standalone post. If you skipped them because they were not applicable, you can simply add `N/A` as an answer and it'll indicate that you've considered the question.
> 
>     * Similarly, the checkmark questions are not answered. It might be that they are losing their effectiveness when you see them tens of times, however we should either discuss a change to it or follow the process in place.
> 
>     * Another consideration is to combine commits for the change from `getColumnIndex` to `getColumnIndexOrThrow` and not create a commit for every single file change. Typically a smaller commit is preferable, but I personally don't see any value in creating 15 commits for this change. Quite the opposite in fact, since the commit messages are only `Update {filename}.java` (in their short form), these commits create quite a lot of noise. Having said all that, this is just my personal take and we don't have any strict processes in place like the other first 2 points I've raised.
> 
> 
> I hope you'll find the review and the feedback useful!

@oguzkocer Thank you for taking time to review and feedback
- I usually fill Regression Notes section.  I was just checking whether the fixes would resolve the issues first.  Updated it now including checklist
-  There were over 150 lines that were hidden in the report.  I had to do one by one with each file update, run lint checks and fix.  I was going to squash them together then I thought someone would say it is a large commit!  Squashed them all into just three commits now.

**Review by ashiagr — APPROVED** [review]
👋 @ravishanker,

I followed the testing instructions and got the build to run successfully.

I found this note in the [Compatibility framework changes (Android 12)](https://developer.android.com/about/versions/12/reference/compat-framework-changes#lock_down_close_system_dialogs).

> For apps targeting Android 12 (API level 31) and higher, the Intent.ACTION_CLOSE_SYSTEM_DIALOGS can't be sent unless the app also has the android.permission.BROADCAST_CLOSE_SYSTEM_DIALOGS permission.

I believe we're good by adding this permission in the manifest.

The resolution to replace `getColumnIndex` with `getColumnIndexOrThrow` looks good to me as well. 👍 

Thank you for fixing these lint issues. 🙏

**Review by oguzkocer — APPROVED** [review]
@ravishanker Thanks for the improvements!

**Comment by ParaskP7** [comment]
👋 @ravishanker @ashiagr !

> I found this note in the [Compatibility framework changes (Android 12)](https://developer.android.com/about/versions/12/reference/compat-framework-changes#lock_down_close_system_dialogs).
> 
> For apps targeting Android 12 (API level 31) and higher, the Intent.ACTION_CLOSE_SYSTEM_DIALOGS can't be sent unless the app also has the android.permission.BROADCAST_CLOSE_SYSTEM_DIALOGS permission.
> 
> I believe we're good by adding this permission in the manifest.

As I am currently trying to upgrade `Gradle & AGP` to `7.4 & 7.1.1` respectively. While doing that, I am getting a new `ProtectedPermissions` Lint error on that `<uses-permission android:name="android.permission.BROADCAST_CLOSE_SYSTEM_DIALOGS" />` permission, telling me that:
```
Permissions with the protection level signature, privileged or signatureOrSystem are only granted to system apps. If an app is a regular non-system app, it will never be able to use these permissions.
```

If I remove this line, instead of suppressing it, I get these other 2 `MissingPermission` Lint errors (`NativeNotificationsUtils` + `ShareAndDismissNotificationReceiver`) that you tried to fix by actually adding this `<uses-permission android:name="android.permission.BROADCAST_CLOSE_SYSTEM_DIALOGS" />` permission.

I am a bit confused and would love your help on that. 🙇

Would you just suppress this `ProtectedPermissions` anyway, even though our app is just a regular non-system app? I actually don't understand why we need `NativeNotificationsUtils` and `ShareAndDismissNotificationReceiver`, which force our app to be considered a system app. Also, I don't know whether us suppressing this Lint error will cause us any trouble down the line, when we start updating to `Android 12` or even `Android 13`. If our app needs to be considered a system app, should we somehow declare that as such, which will most probably resolve the need to suppress this Lint error? 🤔

Any insights about all that and your recommended move going forward?

**Comment by ashiagr** [comment]
👋 Hey @ParaskP7!

Thanks for bringing it to notice. I do not remember seeing this log message:
> Permissions with the protection level signature, privileged or signatureOrSystem are only granted to system apps. If an app is a regular non-system app, it will never be able to use these permissions.

at the time of reviewing the changes in this PR. Maybe it is shown now when you're upgrading `Gradle & AGP` or I might have missed it.

I see your concern about using system permissions. I don't have much context about `NativeNotificationsUtils` and `ShareAndDismissNotificationReceiver` so I briefly checked `ShareAndDismissNotificationReceiver` and it seems it is invoked when we click `Share` button from the `post published notification`. 

Screenshot 2022-02-23 at 3 22 25 PM

[This line](https://github.com/wordpress-mobile/WordPress-Android/blob/a2218f0ea74ff4ee996b6266f09b290c43bc6aa1/WordPress/src/main/java/org/wordpress/android/ui/notifications/ShareAndDismissNotificationReceiver.java#L21): `context.sendBroadcast(new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS))` closes the "system" notification tray when we click "Share" button and shows the app screen behind it otherwise the tray keeps displaying and it is not intuitive that one has to close the tray in order to see the screen behind it.

I think it is not an issue till the app is targeting `Android 11` as the following [apps can't close system dialogs - exception](https://developer.android.com/about/versions/12/behavior-changes-all#close-system-dialogs) is applicable:

> In the following cases, an app can still close system dialogs on Android 12 or higher:

> Your app targets Android 11 or lower. In addition, the user has interacted with a notification, possibly using the notification's [action buttons](https://developer.android.com/training/notify-user/build-notification#Actions), and your app is processing a [service](https://developer.android.com/guide/components/services) or [broadcast receiver](https://developer.android.com/guide/components/broadcasts) in response to that user action.

I also see this note in the same section:

> If your app targets Android 12, you don't need to use ACTION_CLOSE_SYSTEM_DIALOGS in this situation. That's because, if your app calls [startActivity()](https://developer.android.com/reference/android/content/Context#startActivity(android.content.Intent)) while a window is on top of the notification drawer, the system closes the notification drawer automatically.

I'm not quite sure if it is applicable for the exception we're dealing with or the other one when the window is on top of the notification drawer. But I can try that out.  I actually changed `targetSdkVersion` to `31` just to try but see several compilation issues:

```
Manifest merger failed: Apps targeting Android 12 and higher are required to specify an explicit value for android:exported when the corresponding component has an intent filter defined.
```

```
Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
    Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
```

Have we started updating to `Android 12` or is there any branch where these issues are resolved? I can try removing the `ACTION_CLOSE_SYSTEM_DIALOGS` flag and see if the system closes the notification drawer automatically.

For our app targeting `Android 11`, I think we can suppress this `ProtectedPermissions` as the app fulfills the exception criteria.

**Comment by ParaskP7** [comment]
👋 @ashiagr !

Thanks for the response! 🙏 

> at the time of reviewing the changes in this PR. Maybe it is shown now when you're upgrading Gradle & AGP or I might have missed it.

Yes, this is most probably a new error that is now shown with the newest version of `AGP` (the `7.1.1`). I actually didn't even see that when I tried to upgrade to `7.0.4`, so it seems definitely to be new, maybe it was a warning before... 🤷 

> I see your concern about using system permissions. I don't have much context about NativeNotificationsUtils and ShareAndDismissNotificationReceiver so I briefly checked ShareAndDismissNotificationReceiver and it seems it is invoked when we click Share button from the post published notification.
> 
> <IMAGE>
> 
> [This line](https://github.com/wordpress-mobile/WordPress-Android/blob/a2218f0ea74ff4ee996b6266f09b290c43bc6aa1/WordPress/src/main/java/org/wordpress/android/ui/notifications/ShareAndDismissNotificationReceiver.java#L21): context.sendBroadcast(new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS)) closes the "system" notification tray when we click "Share" button and shows the app screen behind it otherwise the tray keeps displaying and it is not intuitive that one has to close the tray in order to see the screen behind it.

Thanks for this thorough investigation Ashita! 🙇 

> I think it is not an issue till the app is targeting Android 11 as the following [apps can't close system dialogs - exception](https://developer.android.com/about/versions/12/behavior-changes-all#close-system-dialogs) is applicable:

Do you maybe mean `Android 12`? 🤔 

> I'm not quite sure if it is applicable for the exception we're dealing with or the other one when the window is on top of the notification drawer. But I can try that out. I actually changed targetSdkVersion to 31 just to try but see several compilation issues:

Expected! 😅

> Have we started updating to Android 12...

As far as I am aware, and since you are the DRI for `Android 12` I don't think we have started the migration just yet.

> ...or is there any branch where these issues are resolved?

Nope... 😊 

> I can try removing the ACTION_CLOSE_SYSTEM_DIALOGS flag and see if the system closes the notification drawer automatically.

That would be a nice try. 🥇 🤞 

> For our app targeting Android 11, I think we can suppress this ProtectedPermissions as the app fulfills the exception criteria.

True, [this is what I did](https://github.com/wordpress-mobile/WordPress-Android/pull/15865/commits/c9702a42c842a3ac80023954a7b4b6b50c6c45e9), at least temporarily.

**Comment by ashiagr** [comment]
> Do you maybe mean Android 12? 🤔

I meant we can close system dialog as we're doing now and suppress `ProtectedPermissions` warning till the app is targeting `Android 11`.

> As far as I am aware, and since you are the DRI for Android 12 I don't think we have started the migration just yet.

Yeah, thank you for the reminder! 🙈 

> [this is what I did](https://github.com/wordpress-mobile/WordPress-Android/pull/15865/commits/c9702a42c842a3ac80023954a7b4b6b50c6c45e9), at least temporarily.

LGTM, 👍 for now.

**Comment by ParaskP7** [comment]
👋 @ashiagr !

> I meant we can close system dialog as we're doing now and suppress ProtectedPermissions warning till the app is targeting Android 11.

Our app is already targeting `Android 11` as we are on `targetSdkVersion = 30` ([Platform Versions & API Levels](https://developer.android.com/guide/topics/manifest/uses-sdk-element#ApiLevels)), isn't that right? Thus, me being confused... 😊 😅 Do you maybe mean `...till the app is targeting Android 12.` instead (that is `targetSdkVersion = 31/32`)? 🤔 

> LGTM, 👍 for now.

Thanks for the confirmation Ashita! 🥇

I'll then proceed with suppressing the error for now. 👍

**Comment by ashiagr** [comment]
Yep, we're targeting `Android 11`,  suppression is valid with the app **still** targeting `Android 11`.  I've avoided using `...till` this time. You've understood it correctly. Let's close this discussion now. 😄
